File: application/src/test/java/com/yahoo/application/container/JDiscContainerSearchTest.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.application.container;

import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

import org.junit.Test;

import com.yahoo.application.Networking;
import com.yahoo.application.container.searchers.AddHitSearcher;
import com.yahoo.component.ComponentSpecification;
import com.yahoo.search.Query;
import com.yahoo.search.Result;

/**
 *
 * @author gjoranv
 * @since 5.1.15
 */
public class JDiscContainerSearchTest {
	@Test
	public void processing_and_rendering_works() throws Exception {
		final String searcherId = AddHitSearcher.class.getName();

		try (JDisc container = containerWithSearch(searcherId)) {
			byte[] rendered = container.search().processAndRender(ComponentSpecification.fromString("mychain"),
					ComponentSpecification.fromString("DefaultRenderer"), new Query(""));
			String renderedAsString = new String(rendered, "utf-8");
			assertThat(renderedAsString, containsString(searcherId));
		}
	}

	@Test
	public void searching_works() throws Exception {
		final String searcherId = AddHitSearcher.class.getName();

		try (JDisc container = containerWithSearch(searcherId)) {
			Search searching = container.search();
			Result result = searching.process(ComponentSpecification.fromString("mychain"), new Query(""));
			String hitTitle = result.hits().get(0).getField("title").toString();
			assertThat(hitTitle, is(searcherId));
		}
	}

	public JDisc containerWithSearch(String searcherId) {
		return JDisc.fromServicesXml("<container version=\"1.0\">" + //
				"<search>" + //
				"<chain id=\"mychain\">" + //
				"<searcher id=\"" + searcherId + "\"/>" + //
				"</chain>" + //
				"</search>" + //
				"</container>", Networking.disable);
	}

	@Test(expected = UnsupportedOperationException.class)
	public void retrieving_search_from_container_without_search_is_illegal() throws Exception {
		try (JDisc container = JDisc.fromServicesXml("<container version=\"1.0\" />", Networking.disable)) {
			container.search(); // throws
		}
	}
}

File: application/src/test/scala/com/yahoo/application/container/JDiscContainerSearchTest.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.application.container

import org.junit.Test
import searchers.AddHitSearcher
import com.yahoo.component.ComponentSpecification
import com.yahoo.search.Query
import org.junit.Assert._
import org.hamcrest.CoreMatchers._
import org.hamcrest.Matchers.containsString;
import JDiscTest.{fromServicesXml, using}


/**
 *
 * @author gjoranv
 * @since 5.1.15
 */
class JDiscContainerSearchTest {

  @Test
  def processing_and_rendering_works() {
    val searcherId = classOf[AddHitSearcher].getName

    using(containerWithSearch(searcherId))
    { container =>
      val rendered = container.search.processAndRender(ComponentSpecification.fromString("mychain"),
        ComponentSpecification.fromString("DefaultRenderer"), new Query(""))
      val renderedAsString = new String(rendered, "utf-8")
      assertThat(renderedAsString, containsString(searcherId))
    }
  }

  @Test
  def searching_works() {
    val searcherId = classOf[AddHitSearcher].getName

    using(containerWithSearch(searcherId))
    { container =>
      val searching = container.search
      val result = searching.process(ComponentSpecification.fromString("mychain"), new Query(""))

      val hitTitle = result.hits().get(0).getField("title").asInstanceOf[String]
      assertThat(hitTitle, is(searcherId))
    }
  }

  def containerWithSearch(searcherId: String) = {
    fromServicesXml(
      <container version="1.0">
        <search>
          <chain id="mychain">
            <searcher id={searcherId}/>
          </chain>
        </search>
      </container>)
  }

  @Test(expected = classOf[UnsupportedOperationException])
  def retrieving_search_from_container_without_search_is_illegal() {
    using(JDiscTest.fromServicesXml(
        <container version="1.0" />
    ))
    { container =>
      container.search  // throws
    }
  }

}
